pallets/gear/src/tests.rs

```diff
@@ -2109,6 +2089,69 @@ fn exit_handle() {
     })
 }
 
+#[test]
+fn calc_gas_spent() {
+    use crate::manager::HandleKind;
+    use tests_btree::{Request, WASM_BINARY};
+
+    init_logger();
+    new_test_ext().execute_with(|| {
+        let initial_balance = BalancesPallet::<Test>::free_balance(USER_1);
+
+        assert_ok!(GearPallet::<Test>::submit_program(
+            Origin::signed(USER_1).into(),
+            WASM_BINARY.to_vec(),
+            DEFAULT_SALT.to_vec(),
+            EMPTY_PAYLOAD.to_vec(),
+            10_000_000,
+            0,
+        ));
+
+        let program_id = utils::get_last_program_id();
+
+        run_to_block(2, None);
+
+        let balance_after_init = BalancesPallet::<Test>::free_balance(USER_1);
+
+        let request = Request::Clear.encode();
+        assert_ok!(GearPallet::<Test>::send_message(
+            Origin::signed(USER_1).into(),
+            program_id,
+            request.clone(),
+            10_000_000,
+            0
+        ));
+
+        run_to_block(3, None);
+
+        let balance_after_handle = BalancesPallet::<Test>::free_balance(USER_1);
+
+        let init_gas_spent = Gear::get_gas_spent(
+            USER_1.into_origin(),
+            HandleKind::Init(WASM_BINARY.to_vec()),
+            EMPTY_PAYLOAD.to_vec(),
+        )
+        .expect("Gas spent for init calculation error");
+
+        assert_eq!(
+            (initial_balance - balance_after_init) as u64,
+            init_gas_spent
+        );
+
+        let handle_gas_spent = Gear::get_gas_spent(
+            USER_1.into_origin(),
+            HandleKind::Handle(program_id),
+            request,
+        )
+        .expect("Gas spent for init calculation error");
+
```

Comment:
Tests for the fact that this call gives the amount of gas which is enough for success execution are cool. But there should be also testing that this call doesn't return some huge amount of gas

Reply:
We can add such kind directly in this test. Need only manually calculate expenses by wat: represent it with mathematical inequality or exact value (it can be changed in future following #658 )

Reply:
I've added a separated test: https://github.com/gear-tech/gear/pull/727/commits/95f78401fea53b1d8b859609f2f13a63c8e29da1